# Data Mining and Visualisation -- Lab 2

In this lab, we will explore some common approaches for Exploratory Data Analysis (EDA).

First off, let's load some useful libraries and functions.

In [19]:
# Import pandas and seaborn, then import stats from scipy
import pandas as pd
import seaborn as sns
from scipy import stats

# Apply the default theme for visualisations
sns.set_theme()

# Make sure that our plots are visualised properly in jupyter
%matplotlib inline



## Descriptive Statistics

There are many ways that we could calculate descriptive statistics in Python. While we could calculate the values ourselves, there are many useful libraries that have done this for us. 

For example, let's go back to the sample that we used in the lecture:

In [20]:
my_sample = pd.Series(data = [1, 3, 6, 7, 9, 10]) 

my_sample

0     1
1     3
2     6
3     7
4     9
5    10
dtype: int64

Say we want to calculate the mean of this dataset. We can do this easily:

In [21]:
my_sample.mean()

6.0

And the standard deviation:

In [22]:
my_sample.std()

3.4641016151377544

This value is the same as what we calculated in the lecture (though with a bit more precision). Note that if you look at the documentation for `std()`, it mentions that it uses N-1 by default (`ddof=1`), calculating the *sample* standard deviation. Should you wish to calculate the *population* standard deviation, we can do so by using the argument `ddof=0` instead.

Let's keep going and find some z-scores:

In [23]:
stats.zscore(my_sample)

0   -1.581139
1   -0.948683
2    0.000000
3    0.316228
4    0.948683
5    1.264911
dtype: float64

So these values are different to the ones we calculated. Why? Because for some strange reason, this function defaults to the *population* standard deviation (`ddof=0`).

Again, we can use the argument `ddof=1` to control for this.

In [24]:
stats.zscore(my_sample, ddof=1)

0   -1.443376
1   -0.866025
2    0.000000
3    0.288675
4    0.866025
5    1.154701
dtype: float64

Similarly, if we wanted to, we could calculate the median of our sample:

In [25]:
my_sample.median()

6.5

And its interquartile range (IQR):

In [26]:
stats.iqr(my_sample)

4.75

## Exploring Your Data

Let's load the titanic dataset and explore it a little.

In [27]:
# Load the 'titanic' example dataset
titanic = pd.read_csv("dataset/titanic.csv")

Often, a great place to start is with the `head()` and `tail()` functions. These functions show the top and bottom rows respectively, and allow us to quickly gather a lot of information about our dataset, including the names of variables, and examples of these values. 

In [28]:
# Show the first five rows of the titanic dataset
titanic.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

By default, both `head()` and `tail()` show five rows, but you can change this by passing in a different `n` argument:

In [29]:
# Show the last ten rows
titanic.tail(n=10)

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
881         0       3    male  33.0      0      0   7.8958        S   Third   
882         0       3  female  22.0      0      0  10.5167        S   Third   
883         0       2    male  28.0      0      0  10.5000        S  Second   
884         0       3    male  25.0      0      0   7.0500        S   Third   
885         0       3  female  39.0      0      5  29.1250        Q   Third   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
881    man        True  NaN  Southampton    no   True  
882  woman       False  NaN  Southampton    no   True  
883    man        True  NaN  Southampton    no   True  
884    man        True  NaN  Southampton    no   True  
885  woman       False  NaN   Queenstown    no  False  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True

We can also use different pandas functions to get more information about our dataset, including...

...`shape` to get the number of rows and columns of our dataset:

In [30]:
titanic.shape

(891, 15)

...`info()` to get a list of the columns and data types:

In [31]:
titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     891 non-null    int64  
 1   pclass       891 non-null    int64  
 2   sex          891 non-null    object 
 3   age          714 non-null    float64
 4   sibsp        891 non-null    int64  
 5   parch        891 non-null    int64  
 6   fare         891 non-null    float64
 7   embarked     889 non-null    object 
 8   class        891 non-null    object 
 9   who          891 non-null    object 
 10  adult_male   891 non-null    bool   
 11  deck         203 non-null    object 
 12  embark_town  889 non-null    object 
 13  alive        891 non-null    object 
 14  alone        891 non-null    bool   
dtypes: bool(2), float64(2), int64(4), object(7)
memory usage: 92.4+ KB


...and `describe()` to gather some descriptive statistics about the numerical variables.

In [32]:
titanic.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

## Cleaning Data



First off, let's modify one of the rows of data, so that it contains a spelling mistake. 

We then count the number of rows for each `embark_town`:

In [33]:
# Modify a row of data to contain a spelling mistake
titanic.loc[(titanic['embark_town'] == 'Cherbourg') & (titanic['sibsp'] > 1) & (titanic['parch'] == 0),'embark_town'] = 'Cherbuorg'

# Count the number of rows for each embark town
titanic['embark_town'].value_counts()

Southampton    644
Cherbourg      167
Queenstown      77
Cherbuorg        1
Name: embark_town, dtype: int64

Our spelling mistake shows up! We have 167 instances of 'Cherbourg' and 1 instance of 'Cherbuorg'. Let's fix that as if we had only just found it:

In [34]:
# Correct the spelling mistake
titanic.loc[titanic['embark_town'] == 'Cherbuorg','embark_town'] = 'Cherbourg'

# Count the number of rows with each embark town
titanic['embark_town'].value_counts()

Southampton    644
Cherbourg      168
Queenstown      77
Name: embark_town, dtype: int64

And the spelling mistake is fixed!

Next up, let's have a look for some missing data. We filter our titanic dataset to only include rows where `embark_town` is missing:

In [39]:
# Show the rows of data where embark_town is NaN

titanic[titanic.embark_town.isnull()]

survived  pclass     sex   age  sibsp  parch  fare embarked  class  \
61          1       1  female  38.0      0      0  80.0      NaN  First   
829         1       1  female  62.0      0      0  80.0      NaN  First   

       who  adult_male deck embark_town alive  alone  
61   woman       False    B         NaN   yes   True  
829  woman       False    B         NaN   yes   True

There are a number of options available within pandas for [working with](https://pandas.pydata.org/docs/user_guide/missing_data.html#working-with-missing-data), and [filling in](https://pandas.pydata.org/docs/user_guide/missing_data.html#filling-missing-data) missing data. 

If you have time, or if you're ever facing missing data within your dataset, having a read of the pandas documentation can be very useful for exploring some of the options and techniques in more detail!

## More Advanced Querying

We'll now show a few examples of useful queries to get you started in exploring datasets.

First off, we can obtain a subset of our data, using conditional logic. For example, if we wanted to get a subset of the data only where `embark_town` == Cherbourg, we can do so as follows:

In [18]:
# Filter based on a condition
titanic[titanic["embark_town"] == "Cherbourg"]

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
1           1       1  female  38.0      1      0  71.2833        C   First   
9           1       2  female  14.0      1      0  30.0708        C  Second   
19          1       3  female   NaN      0      0   7.2250        C   Third   
26          0       3    male   NaN      0      0   7.2250        C   Third   
30          0       1    male  40.0      0      0  27.7208        C   First   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
866         1       2  female  27.0      1      0  13.8583        C  Second   
874         1       2  female  28.0      1      0  24.0000        C  Second   
875         1       3  female  15.0      0      0   7.2250        C   Third   
879         1       1  female  56.0      0      1  83.1583        C   First   
889         1       1    male  26.0      0      0  30.0000        C   First   

       who  adult_male deck embark_town alive  alone  
1    woman       False    C   Cherbourg   yes  False  
9    child       False  NaN   Cherbourg   yes  False  
19   woman       False  NaN   Cherbourg   yes   True  
26     man        True  NaN   Cherbourg    no   True  
30     man        True  NaN   Cherbourg    no   True  
..     ...         ...  ...         ...   ...    ...  
866  woman       False  NaN   Cherbourg   yes  False  
874  woman       False  NaN   Cherbourg   yes  False  
875  child       False  NaN   Cherbourg   yes   True  
879  woman       False    C   Cherbourg   yes  False  
889    man        True    C   Cherbourg   yes   True  

[168 rows x 15 columns]

We can then use that subset to perform more segmented analysis:

In [19]:
titanic[titanic["embark_town"] == "Cherbourg"].age.mean()

30.81476923076923

However, let's say that we want to do that for *each* `embark_town` values, and not just Cherbourg. 

To do this, we can use the *split-apply-combine* paradigm that we discussed in the lecture to perform these analyses for *each* of our groups:

In [20]:
# Group by and count each value
titanic.groupby("embark_town").age.mean()

embark_town
Cherbourg      30.814769
Queenstown     28.089286
Southampton    29.445397
Name: age, dtype: float64

Here, we have only scratched the surface of what is possible within pandas for exploratory data analysis. You're encouraged to go find some other examples (tutorials, uploaded Kaggle code, documentation) to further learn how pandas can help you to shape and analyse your data.

## Exploring a Dataset of Your Choosing

In the previous lab, you were tasked with finding some interesting datasets from [Kaggle](https://www.kaggle.com/) or other data repositories. 

Try loading the dataset into python and have a go at exploring it in more detail. Can you find anything interesting about the data?